### PR TITLE
[HOTFIX] JDBC completions. fix cursor position

### PR DESCRIPTION
**Pull request by tinkoff-dwh**
### What is this PR for?
If text of paragraph contains interpreter name then parser works incorrect (cursor position  incorrect). 

### What type of PR is it?
Bug Fix

### Screenshots (if appropriate)
before
before
after
after

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java [hunk 1]
```diff
@@ -291,7 +314,10 @@ public Interpreter getCurrentRepl() {
 
     String replName = getRequiredReplName(buffer);
     if (replName != null && cursor > replName.length()) {
-      cursor -= replName.length() + 1;
+      Matcher m = Pattern.compile(String.format("%%%s\\s*", replName)).matcher(buffer);
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: I'm not sure we should have another way to match here? should we correct/reuse the logic in L307, or getRequiredReplName etc?

Comment (tinkoff-dwh) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: @felixcheung 
L307 logic to find replName. Now problem with whitespace character. For example
```
%replName\n
\n
\n
text<cursor>
```
result **cursor=7** but method **getScriptBody** return ```text```

Need count all characters (for correct value of cursor).

**Comment by tinkoff-dwh** [comment]
Ready to review

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java (hunk 1 above)
Comment (tinkoff-dwh) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: @felixcheung ?

Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: I see, should just substring `%replName` and then look in the rest of the string but account for whitespaces?

I'm a bit worry about more issues with parsing `%something` - we have seen quite a bit in the past

**Comment by tinkoff-dwh** [comment]
@felixcheung 
another solution to count whitespaces
CI green https://travis-ci.org/tinkoff-dwh/zeppelin/builds/257656645

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java [hunk 2]
```diff
@@ -179,7 +179,8 @@ public int complete(String buffer, int cursor, List<InterpreterCompletion> candi
 
 
   private static void fillTableNames(String schema, DatabaseMetaData meta, Set<String> tables) {
-    try (ResultSet tbls = meta.getTables(schema, schema, "%", null)) {
+    try (ResultSet tbls = meta.getTables(schema, schema, "%",
+        new String[]{"TABLE", "VIEW", "ALIAS", "SYNONYM", "GLOBAL TEMPORARY", "LOCAL TEMPORARY"})) {
```
Comment (felixcheung) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: can you elaborate on this change?

Comment (felixcheung) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: "SYNONYM", "GLOBAL TEMPORARY", "LOCAL TEMPORARY" might not be globally applicable to all JDBC sources

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java [hunk 3]
```diff
@@ -291,7 +314,8 @@ public Interpreter getCurrentRepl() {
 
     String replName = getRequiredReplName(buffer);
     if (replName != null && cursor > replName.length()) {
-      cursor -= replName.length() + 1;
+      String body = buffer.substring(replName.length() + 1);
+      cursor -= replName.length() + 1 + body.indexOf(body.trim());
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: I think maybe the trim() on body should go before the substring? say for example
```
  %foo    select


* from table
```

if there's whitespace before replName

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java (hunk 2 above)
Comment (tinkoff-dwh) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: it is filter of returned "entities" in the method `getTables`. If value is null then returns all types. for example, returning the name of the indexes from tables (postrges)

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java (hunk 3 above)
Comment (tinkoff-dwh) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: then cursor would be incorrect

**Comment by tinkoff-dwh** [comment]
@felixcheung 
done

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java (hunk 2 above)
Comment (felixcheung) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: I'm afraid this varies depending on the source, for instance it could return quite a different list for SQL server
https://docs.microsoft.com/en-us/sql/connect/jdbc/reference/gettables-method-sqlserverdatabasemetadata

Comment (felixcheung) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: and perhaps this should be a different PR?

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/ParagraphTest.java [hunk 4]
```diff
@@ -228,8 +226,34 @@ public void returnUnchangedResultsWithDifferentUser() throws Throwable {
...
+        Triple.of("   %jdbc schema.", 16, 7),
+        Triple.of(" \n%jdbc schema.", 15, 7),
+        Triple.of("%jdbc schema.table.  ", 19, 13),
+        Triple.of("%jdbc schema.\n\n", 13, 7),
+        Triple.of("  %jdbc schema.tab\n\n", 18, 10),
+        Triple.of("  \n%jdbc schema.\n \n", 16, 7),
+        Triple.of("  \n%jdbc schema.\n \n", 16, 7),
+        Triple.of("  \n%jdbc\n\n schema\n \n", 17, 6),
+        Triple.of("%another\n\n schema.", 18, 7),
+        Triple.of("\n\n schema.", 10, 7),
```
Comment (felixcheung) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/ParagraphTest.java]: add more tests for, for example
```
  \n   %jdbc
\n   %jdbc
%jdbc  \n  schema
%jdbc  \n  \n   schema
```

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java (hunk 2 above)
Comment (tinkoff-dwh) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: SQL server have same types (TABLE, VIEW) 
https://github.com/Microsoft/mssql-jdbc/blob/master/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDatabaseMetaData.java#L1268

**Comment by tinkoff-dwh** [comment]
@felixcheung 
done

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java (hunk 2 above)
Comment (felixcheung) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: I'm not if it is though. in your link, there is no 'SYNONYM'

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java [hunk 5]
```diff
@@ -306,6 +329,23 @@ public Interpreter getCurrentRepl() {
...
+  public int calculateCursorPosition(String buffer, int cursor) {
+    int countWhitespacesAtStart = buffer.indexOf(buffer.trim());
+    if (countWhitespacesAtStart > 0) {
+      cursor -= countWhitespacesAtStart;
+    }
+
+    buffer = buffer.trim();
+
+    String replName = getRequiredReplName(buffer);
+    if (replName != null && cursor > replName.length()) {
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: nit: there's a fair bit of overlap here. here and `completion()`, `buffer.trim()` and `getRequireReplName(buffer)` is called x2

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java (hunk 2 above)
Comment (tinkoff-dwh) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: it's not a problem

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java (hunk 5 above)
Comment (masyan) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: called x5 )
String replName = getRequiredReplName(buffer)  x1
String body = getScriptBody(buffer) (getRequiredReplName, String trimmed = text.trim())_x2
calculateCursorPosition ( buffer = buffer.trim(), getRequiredReplName)

Comment (tinkoff-dwh) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: called x5 )
String replName = getRequiredReplName(buffer) x1
String body = getScriptBody(buffer) (getRequiredReplName, String trimmed = text.trim())_x2
calculateCursorPosition ( buffer = buffer.trim(), getRequiredReplName)

**Comment by tinkoff-dwh** [comment]
@felixcheung 
fixed

CI RED because
 W: GPG error: http://cran.rstudio.com/bin/linux/ubuntu precise/ Release: The following signatures were invalid: BADSIG 51716619E084DAB9 Michael Rutter <marutter@gmail.com>

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by tinkoff-dwh** [comment]
@felixcheung 
thx for review. could you initiate merge if all ok?

**Comment by tinkoff-dwh** [comment]
fix is very important, because auto complete not works. And if you have many shemas and tables it it hell.

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by Leemoonsoo** [comment]
Thanks, @tinkoff-dwh for the important fix.

I think different reserved words (like "TABLE", "VIEW", "ALIAS", "SYNONYM", "GLOBAL TEMPORARY", "LOCAL TEMPORARY") on different database can be handled in future PRs.

LGTM

**Comment by Leemoonsoo** [comment]
Merge to master if no further comments

**Comment by tinkoff-dwh** [comment]
@Leemoonsoo @felixcheung 
pls, merge PR
